src/tiled/scriptfileformatwrappers.cpp

```diff
@@ -0,0 +1,128 @@
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "scriptfileformatwrappers.h"
+
+#include "editablemap.h"
+#include "editabletileset.h"
+#include "mapformat.h"
+#include "scriptedfileformat.h"
+#include "scriptmanager.h"
+#include "tilesetformat.h"
+
+#include <QCoreApplication>
+#include <QQmlEngine>
+
+namespace Tiled {
+
+ScriptTilesetFormatWrapper::ScriptTilesetFormatWrapper(TilesetFormat* format, QObject *parent)
+    : QObject(parent)
+    , mFormat(format)
+{}
+
+QString ScriptTilesetFormatWrapper::name() const
+{
+    if (auto scriptedFormat = qobject_cast<ScriptedTilesetFormat*>(mFormat))
+        return scriptedFormat->name();
+    else
+        return mFormat->shortName();
+}
+
+QString ScriptTilesetFormatWrapper::extension() const
+{
+    if (auto scriptedFormat = qobject_cast<ScriptedTilesetFormat*>(mFormat))
+        return scriptedFormat->extension();
+    else
+        return QString();
+}
+
+EditableTileset *ScriptTilesetFormatWrapper::read(QString &filename)
+{
+    auto tileset = mFormat->read(filename);
+    auto editable = new EditableTileset(tileset.data());
+    QQmlEngine::setObjectOwnership(editable, QQmlEngine::JavaScriptOwnership);
+    return editable;
+}
+
+QString ScriptTilesetFormatWrapper::write(EditableTileset *editable, QString &filename)
+{
+    auto tileset = editable->tileset();
+    if (mFormat->write(*tileset, filename))
+        return QString();
+    else
+        return mFormat->errorString();
+}
+
+bool ScriptTilesetFormatWrapper::supportsFile(QString &filename) const
+{
+    return mFormat->supportsFile(filename);
+}
+
+ScriptMapFormatWrapper::ScriptMapFormatWrapper()
+{
+    auto message = QCoreApplication::translate("Script Errors",
+                                               "MapFormatWrapper cannot be constructed");
+    ScriptManager::instance().throwError(message);
+}
+
+ScriptMapFormatWrapper::ScriptMapFormatWrapper(MapFormat *format, QObject *parent)
+    : QObject(parent)
+    , mFormat(format)
+{}
+
+QString ScriptMapFormatWrapper::name() const
+{
+    if (auto scriptedFormat = qobject_cast<ScriptedMapFormat*>(mFormat))
+        return scriptedFormat->name();
+    else
+        return mFormat->shortName();
+}
+
+QString ScriptMapFormatWrapper::extension() const
+{
+    if (auto scriptedFormat = qobject_cast<ScriptedMapFormat*>(mFormat))
+        return scriptedFormat->extension();
+    else
+        return QString();
+}
+
+EditableMap *ScriptMapFormatWrapper::read(QString &filename)
+{
+    auto map = mFormat->read(filename);
+    auto editable = new EditableMap(std::move(map));
+    QQmlEngine::setObjectOwnership(editable, QQmlEngine::JavaScriptOwnership);
+    return editable;
+}
+
+QString ScriptMapFormatWrapper::write(EditableMap *editable, QString &filename)
+{
+    auto map = editable->map();
+    if (mFormat->write(map, filename))
+        return QString();
+    else
+        return mFormat->errorString();
```

Comment:
I wonder if it's better to throw a script error here.

Reply:
See the comment above; we expect errors to be returned as a string in user-supplied write functions, so I made the wrapper do the same.

We can definitely throw an error here instead, just wanted to try to keep things consistent with the user-supplied objects.

Reply:
Another option is to only return a string if the backing TilesetFormat is castable to ScriptedTilesetFormat (similarly for MapFormat), and otherwise throw an error?

Reply:
Let's not implement different behaviors depending on the implementation of the file format. I think throwing an error yields a better default behavior, which can be tweaked by using `try` and `catch` if needed.